Kuzzle/Offline/OfflineManager.cs

```diff
@@ -0,0 +1,200 @@
+    /// If set to -1, the size is unlimited.
+    /// </summary>
+    public override int MaxQueueSize {
+      get { return maxQueueSize; }
+      set { maxQueueSize = value < 0 ? -1 : value; }
+    }
+
+    /// <summary>
+    /// The minimum duration of a Token after refresh.
+    /// If set to -1 the SDK does not refresh the token automaticaly.
+    /// </summary>
+    public override int RefreshedTokenDuration {
+      get { return refreshedTokenDuration; }
+      set { refreshedTokenDuration = value < 0 ? -1 : value; }
+    }
+
+    /// <summary>
+    /// The minimum duration of a Token before being automaticaly refreshed.
+    /// If set to -1 the SDK does not refresh the token automaticaly.
+    /// </summary>
+    public override int MinTokenDuration {
+      get { return minTokenDuration; }
+      set { minTokenDuration = value < 0 ? -1 : value; }
+    }
+
+    /// <summary>
+    /// The maximum delay between two requests to be replayed
+    /// </summary>
+    public override int MaxRequestDelay {
+      get { return maxRequestDelay; }
+      set { maxRequestDelay = value < 0 ? 0 : value; }
+    }
+
+    public override Func<JObject, bool> QueueFilter {
+      get { return queueFilter; }
+      set { queueFilter = value ?? ((obj) => true); }
+    }
+
+    /// <summary>
+    /// Queue requests when network is down,
+    /// and automatically replay them when the SDK successfully reconnects.
+    /// </summary>
+    public override bool AutoRecover { get; set; } = true;
+
+    internal OfflineManager(AbstractProtocol networkProtocol, IKuzzle kuzzle) {
+      this.NetworkProtocol = networkProtocol;
+      networkProtocol.StateChanged += this.StateChangeListener;
+      this.kuzzle = kuzzle;
+      InitComponents();
+      this.kuzzle.GetEventHandler().UserLoggedIn += this.OnUserLoggedIn;
+    }
+
+    internal virtual void InitComponents() {
+      QueryReplayer = new QueryReplayer(this, kuzzle);
+      SubscriptionRecoverer = new SubscriptionRecoverer(this, kuzzle);
+      TokenVerifier = new TokenVerifier(this, kuzzle);
+    }
+
+    /// <summary>
+    /// Gets the network protocol.
+    /// </summary>
+    /// <returns>The network protocol.</returns>
+    internal override AbstractProtocol NetworkProtocol { get; set; }
+
+    /// <summary>
+    /// Gets the query replayer.
+    /// </summary>
+    /// <returns>The query replayer.</returns>
+    internal override IQueryReplayer QueryReplayer { get; set; }
+
+    /// <summary>
+    /// Gets the token verifier.
+    /// </summary>
+    /// <returns>The token verifier.</returns>
+    internal override ITokenVerifier TokenVerifier { get; set; }
+
+    /// <summary>
+    /// This will check the token validity,
+    /// and chose what to do before replaying the Queue
+    /// </summary>
+    internal override async Task Recover() {
+      if (!AutoRecover) return;
+
+      if (await TokenVerifier.IsTokenValid()) {
+        QueryReplayer.ReplayQueries();
+        QueryReplayer.Lock = false;
```

Comment:
We can only unlock the queue after the query locking it (either a login or a logout) has been settled (either successfully or not). Until then, we cannot know with what credentials requests must be executed (the queued login might fail for instance).